tools/clang/lib/AST/HlslTypes.cpp

```diff
@@ -103,31 +103,26 @@ bool IsHLSLNumericOrAggregateOfNumericType(clang::QualType type) {
          BuiltinTy->getKind() != BuiltinType::Kind::Char_S;
 }
 
-bool IsHLSLNumericUserDefinedType(clang::QualType type) {
-  const clang::Type *Ty = type.getCanonicalType().getTypePtr();
+// In some cases we need record types that are annotatable and trivially
+// copyable from outside the shader. This excludes resource types which may be
+// trivially copyable inside the shader, and builtin matrix and vector types
+// which can't be annotated. But includes UDTs of trivially copyable data and
+// the builtin trivially copyable raytracing structs.
+bool IsHLSLCopyableAnnotatableRecord(clang::QualType QT) {
+  const clang::Type *Ty = QT.getCanonicalType().getTypePtr();
   if (const RecordType *RT = dyn_cast<RecordType>(Ty)) {
     const RecordDecl *RD = RT->getDecl();
-    if (!IsUserDefinedRecordType(type))
+    if (!IsUserDefinedRecordType(QT))
       return false;
-    for (auto member : RD->fields()) {
-      if (!IsHLSLNumericOrAggregateOfNumericType(member->getType()))
+    for (auto Member : RD->fields()) {
+      if (!IsHLSLNumericOrAggregateOfNumericType(Member->getType()))
         return false;
     }
     return true;
   }
   return false;
 }
 
-// In some cases we need record types that are annotatable and trivially
-// copyable from outside the shader. This excludes resource types which may be
-// trivially copyable inside the shader, and builtin matrix and vector types
-// which can't be annotated. But includes UDTs of trivially copyable data and
-// the builtin trivially copyable raytracing structs.
-bool IsHLSLCopyableAnnotatableRecord(clang::QualType QT) {
-  return IsHLSLNumericUserDefinedType(QT) ||
-         IsHLSLBuiltinRayAttributeStruct(QT);
```

Comment:
The cumulative effect of these changes is that in instances where `IsHLSLNumericUserDefinedType` would have failed and `IsHLSLBuiltinRayAttributeStruct` would have succeeded, `IsHLSLCopyableAnnotatableRecord` will fail. That doesn't seem consistent with the description of treating DXR struct types the same as UDTs. Is that just not possible because all DXR types will always pass `IsHLSLCopyableAnnotatableRecord` anyway? The comment describing it as passing raytracing structs hasn't changed.

Reply:
See comment below, this is just a collapse of two functions into one, which are identical in behavior after the `IsUserDefinedRecordType` change was made.

There is no longer a case where `IsHLSLNumericUserDefinedType` would fail and `IsHLSLBuiltinRayAttributeStruct` would have succeeded due to that change in `IsUserDefinedRecordType`.  With that change, if `IsHLSLBuiltinRayAttributeStruct` returns true, then `IsHLSLNumericUserDefinedType` would also return true.

> Is that just not possible because all DXR types will always pass IsHLSLCopyableAnnotatableRecord anyway?

Yes, that's right.

> The comment describing it as passing raytracing structs hasn't changed.

Because it still does return true for DXR structs.

Reply:
What I missed was that one of the calls in IsHLSLCopyableANnotatableRecord was to the IsUserDefinedRecordType changed below. This makes sense now. Thanks for the clarification!